### DISCUSSION TITLE
Reproducing URQL issues

### DISCUSSION THREAD

**Discussion by erquhart** [urql-graphql/urql#3269] [Q&A]
I love this library *so much*. About 97% of the time. The other 3% poses a very real risk of physical harm to my poor, defenseless MacBook.

The number one blocker to getting resolution to issues here, in my experience, is that fully reproducing a unique issue feels like a pretty solid undertaking. Do you as maintainers have any kind of starter that you use for reproducing, or any pointers on how you go about it repros?

I'm specifically working with React Native, so I'd probably have to make an Expo Snack, and then there's mocking out a suitable backend. Any help is greatly appreciated.

**Comment by kitten** [comment]
In _most_ cases, there shouldn't be any need to start up a React Native application, even if the issue lies with it. The exception are some unique `Promise` timing issues (React Native seems to have a bit of a special behaviour here in terms of tick-batching), and some `fetch` differences. That said, most of these are outright "app-busting errors", and hence pretty easy to track down, unless they're polyfill issues.

In any case, we've got a very large library of examples: https://github.com/urql-graphql/urql/tree/main/examples

These are all in a separate pnpm workspace, so when I'm working on issues, I reproduce them or create an environment in an example, then `pnpm link` the built packages over. This is sufficient to test issues and fix resolution.

Similarly, when someone submits a reproduction, my aim is usually to clone/download them, link some built packages over when the `debugExchange` and logging isn't enough, then observe internal behaviour after a `pnpm link`.

I haven't yet discovered how the new CodeSandbox handles downloads 🤷 It's a bit of a shame, because most of the time, I'm not interested in the slightest in debugging _inside_ CodeSandbox.

Anyway, the examples folder has a bunch of criteria for us:
- Each example should show a unique feature
- Each example should be an effective demo/explanation of a unique feature
- Each example should be standalone runnable
  - the API must be stabily deployed or included in the example
  - all dependencies must be specified local-first, i.e. copying the example folder won't break the example

**Reply by erquhart** [reply]
Super helpful, thank you. Do you mock the graphql api somehow for repro purposes?
